### PR TITLE
Snooze state display in UI

### LINKED ISSUE #1575: GUI: show cluster snooze status
Follow-up of #1572 

- web API: when we are starting to snooze, send an event over the websocket connection
- in the GUI: add snooze/un-snoozing/ready cluster status
    - show this in the "connected > cluster info" button, especially to give feedback when the cluster is un-snoozing (animate that state?)

### PR DESCRIPTION
**Pull request by matbryan52**
Had a stab at implementing the cluster snooze state display to respond to #1575 .

image

It is functional but I actually don't know how best to implement the Python side (ironically), as I think this is a unique case where the Python side autonomously triggers a change in the web interface rather than handling a request through the API. In particular the `ExecutorState`, which is doing the snoozing but has no knowledge of the `SharedState` or `EventRegistry` cannot send messages to the web interface. It can be hacked-in but this creates some circular references and weakens the separation of concerns.

Rather than just sit on the proto I thought I would put it up for comments and/or advice!

Right now this only implements the transition to "snoozed" and back to "connected". It does not yet implement the "unsnoozing" in-progress state.

Fixes #1575 

## Contributor Checklist:

* [x] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [ ] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [ ] I have added/updated documentation for all user-facing changes
* [ ] I have added/updated test cases
* [ ] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if changes were made to the GUI)

## Reviewer Checklist:

* [ ] `/azp run libertem.libertem-data` passed
* [ ] No import of GPL code from MIT code

### CODE REVIEW

**Review by sk1p — COMMENTED** [review]
Thank you for getting this started. The TS changes are pretty much what I would have written :+1:

> I think this is a unique case where the Python side autonomously triggers a change in the web interface rather than handling a request through the API.

Indeed - a similar case are the events emitted from running jobs, and their corresponding progress messages. But it's true that this is initiated by a request, and thus comes through a handler that has access to the whole server-side "state tree".

> In particular the ExecutorState, which is doing the snoozing but has no knowledge of the SharedState or EventRegistry cannot send messages to the web interface. It can be hacked-in but this creates some circular references and weakens the separation of concerns.

So, the cycle would be: `ExecutorState` refs `EventRegistry` refs `ResultEventHandler`s refs (`SharedState` refs **`ExecutorState`**) + **`EventRegistry`**. The `EventRegistry <-> ResultEventHandler` cycle is already there, but IMHO okay, as these classes are closely related and have high cohesion.

The "concern weakening" would be that with the change, the `EventRegistry` is no longer only bound to all the HTTP handlers, but also becomes known to internal components (like `ExecutorState`), which previously only kept state around. Is this your line of thinking?

Right now, the `EventRegistry` is quite simple - it's essentially a broadcasting event bus, delivering the messages to all of the websocket connections.

We could think about adding a secondary event bus, which is replicated to the websocket connections via the `EventRegistry`, but lives on a higher level of the whole object tree - it could be injected into both the `EventRegistry` and the `ExecutorState`:

```python
event_bus = EventBus()  # not attached to the name, feel free to bikeshed
event_registry = EventRegistry(event_bus=event_bus)
executor_state = ExecutorState(..., event_bus=event_bus)

def snooze(self):
    # ...
    self.event_bus.send(some_msg)
```

Then, there would need to be a kind of `MessagePump` from the internal event bus to the websocket connection for forwarding the messages. I don't know if this event bus would even need to be bi-directional - probably an unidirectional forwarding to the websockets would be enough for now (pattern: multiple producers, single consumer).

This event bus can then be injected into all classes that somehow generate events, even those that don't necessarily have access to the `SharedState`. (Aside: as I just built a similar thing in another context: in Python it might make sense to use a `queue.Queue` internally, as it is thread-safe but still can be used in an async context via `loop.run_in_executor(...)`. Specifically, `asyncio.Queue` is _not_ thread-safe, and can thus also not bridge different async loops (unless you lug around both the queue and the associated loop, and use `loop.call_soon_threadsafe`, ugh...))

Another topic is that the `Message` class needs to know about the whole `SharedState` right now - I think that doesn't need to be the case. `Message.state` is not used very widely (`start_job` / `finish_job`), and `Message` is mostly meant to keep the message serialization close together. If you like, I can take care of this aspect, as that's just a churny refactoring job, not really related to this PR.

With these changes, the `EventRegistry` itself doesn't change and isn't coupled to the `ExecutorState`, there's just an additional component attached to the `EventRegistry` that pumps asynchronous events from wherever in the system to the websocket connections. There is no additional cycle added, as `EventRegistry` and `ExecutorState` are only indirectly connected via the added event bus.

(Is that correct? Something like: `ExecutorState` refs `EventBus`, `MessagePump` refs `EventBus`, `MessagePump` refs `EventRegistry`?)

I hope this makes some kind of sense - I have to think a bit about if the added complexity is worth it, or if there is maybe an easier solution.

File: client/src/clusterStatus/components/Cluster.tsx [hunk 1]
```diff
@@ -23,21 +23,28 @@ type MergedProps = ReturnType<typeof mapStateToProps>;
...
             setStatus(clusterConnection.status)
             setDisable(false)
-            if (clusterConnection.status === "connected") {
+            if (channelStatus === "ready") {
                 setColor(ColorType.blue)
+                setStatus("connected")
+            }
+            else if (channelStatus === "snoozed") {
+                setColor(ColorType.red)
+                setStatus("snoozed")
```
[the remark this answers is not in the document]
Comment (sk1p) [reply client/src/clusterStatus/components/Cluster.tsx:44]: Minor comment on visuals: having the button red might be a bit crass - does that state warrant a "caution" signal? Maybe grey plus the "broken chain" icon would work?

image

[the remark this answers is not in the document]
Comment (matbryan52) [reply client/src/clusterStatus/components/Cluster.tsx:44]: > Minor comment on visuals: having the button red might be a bit crass 

Definitely agreed, I was mainly seeing if I could implement the TS part more than making it nice, for now.

**Comment by matbryan52** [comment]
I've added the "unsnoozing" state between "snoozed" and "ready".

However it doesn't work fully right now I think because executor creation blocks the event loop, meaning the `unsnooze` message does not arrive until after the executor is back up again (again #1577)..

I've also noticed that opening the cluster info modal causes the cluster to unsnooze (which could take a while). It's an edge case but I will look into how hard it will be to avoid this.

**Comment by sk1p** [comment]
> I've added the "unsnoozing" state between "snoozed" and "ready".

Great!

> However it doesn't work fully right now I think because executor creation blocks the event loop, meaning the `unsnooze` message does not arrive until after the executor is back up again (again #1577)..

Yup, it's basically this issue we need to tackle in a different way:

https://github.com/LiberTEM/LiberTEM/blob/5d4d2de2256113b3760fabede1a316716fadaa6b/src/libertem/web/server.py#L244-L252

I'll try to find some time to look at this later today.

> I've also noticed that opening the cluster info modal causes the cluster to unsnooze (which could take a while). It's an edge case but I will look into how hard it will be to avoid this.

Probably, the "cluster info" would need to be cached, such that it is available without a running executor.

**Comment by sk1p** [comment]
> However it doesn't work fully right now I think because executor creation blocks the event loop, meaning the `unsnooze` message does not arrive until after the executor is back up again (again #1577)..

This should now be fixed, by making the necessary methods `async`. What I see now on the websocket:

image

I think this did the job, let's see if this passes CI cleanly (older Python/distributed versions might disagree with me...)

**Comment by sk1p** [comment]
CI failure seems unrelated to the changes, caused by `distributed==2024.2.1` (works fine with `distributed==2024.2.0`)

**Comment by sk1p** [comment]
> CI failure seems unrelated to the changes, caused by `distributed==2024.2.1` (works fine with `distributed==2024.2.0`)

Tracked down somehow to https://github.com/dask/dask/pull/10883

**Comment by sk1p** [comment]
> > CI failure seems unrelated to the changes, caused by `distributed==2024.2.1` (works fine with `distributed==2024.2.0`)
> 
> Tracked down somehow to [dask/dask#10883](https://github.com/dask/dask/pull/10883)

... and also, if I `log.debug("%r", params_fut)` in `_get_future`, the issue goes away (wtf?). If I just manually call `repr(params_fut)` in there, the bug is still triggered.

**Comment by sk1p** [comment]
Sorry for the noise, the event loop issues on Python 3.7 should be ironed out now - and another workaround for dask has been cherry-picked into #1604 and has already already merged. Possibly these issues need to be revisited for Python 3.8 in #1603.

@matbryan52 feel free to take over again!

**Comment by matbryan52** [comment]
> Sorry for the noise, the event loop issues on Python 3.7 should be ironed out now - and another workaround for dask has been cherry-picked into #1604 and has already already merged. Possibly these issues need to be revisited for Python 3.8 in #1603.
> 
> @matbryan52 feel free to take over again!

No worries, and thanks for reviewing / adding the async methods.

For me this is good to go, I've included a rebuilt client now.

**Pull request merged**